test/test_session.py

```diff
@@ -173,6 +176,36 @@ def _test_ops(self, client, *ops):
                         "%s did not return implicit session to pool" % (f.__name__,),
                     )
 
+    def test_implicit_sessions_checkout(self):
+        # "To confirm that implicit sessions only allocate their server session after a
+        # successful connection checkout" test from Driver Sessions Spec.
+        listener = EventListener()
+        client = rs_or_single_client(event_listeners=[listener], maxPoolSize=1, retryWrites=True)
+
+        ops = [
+            (client.db.test.find_one, [{"_id": 1}]),
+            (client.db.test.delete_one, [{}]),
+            (client.db.test.update_one, [{}, {"$set": {"x": 2}}]),
+            (client.db.test.bulk_write, [[UpdateOne({}, {"$set": {"x": 2}})]]),
+            (client.db.test.find_one_and_delete, [{}]),
+            (client.db.test.find_one_and_update, [{}, {"$set": {"x": 1}}]),
+            (client.db.test.find_one_and_replace, [{}, {}]),
+            (client.db.test.find, [{}]),
+        ]
+        threads = []
+        listener.results.clear()
+        for op, args in ops:
+            threads.append(threading.Thread(target=op, args=typing.cast(typing.Iterable, args)))
```

Comment:
What's the `typing.cast(typing.Iterable` here? Should we type hint `ops` instead?